---
title: "Correlation Modified by HIV Status"
author: "Guannan Shen"
date: "April 1, 2019"
output: 
  html_document:
    number_sections: yes
    theme: united
    toc: yes
    toc_depth: 5
    toc_float: yes
  word_document:
    toc: yes
    toc_depth: '5'
  pdf_document:
    latex_engine: lualatex
    number_sections: yes
    toc: yes
    toc_depth: 5
---

```{r setup, include=FALSE, cache = FALSE}
require("knitr")
opts_chunk$set(tidy.opts=list(width.cutoff=60),tidy=TRUE)
opts_chunk$set(engine = "R")
knitr::opts_chunk$set(echo = T)
knitr::opts_chunk$set(message = F)
knitr::opts_chunk$set(warning = F)
## setting wd in DELL
## opts_knit$set(root.dir = "~/Stats/CIDA_OMICs/CIDA_OMICS/7659Stats_Genetics/HW5/")
## setting working directory in asus 
## opts_knit$set(root.dir = "C:/Users/hithr/Documents/Stats/gitlab/Cario_RNASeq_Microbiom_Inte/DataRaw/") 
## setting working directory in ubuntu
opts_knit$set(root.dir = "~/Documents/gitlab/Cario_RNASeq_Microbiom_Inte/DataProcessed/Vgenes/")
                                                 
## cache = F, if cache = T, will not revaluate code chunk everytime
## double or more space to insert a line break
```


```{r libs, echo = FALSE}
######################
## Set up workspace
######################
rm(list = ls())
library(knitr)
library(tidyverse)
library(magrittr)
library(stats)
library(readxl)
library(openxlsx)
# multiple testing
library(mutoss)
options(stringsAsFactors = F)
options(dplyr.width = Inf)
getwd()
## not in function
'%nin%' <- Negate('%in%')

# ######## clean memory ######################
# rm(list = ls())
# gc()
# is(dds)
# slotNames(dds)

```

## Data


```{r importdata, echo = FALSE}
############### all V genes #####################
# rnaseq
rnaseq <- read_excel( "vgenes_rnaseq.xlsx")
head(rnaseq)
dim(rnaseq)
# rnaseq[rnaseq$HIV == "no", 1:45]
# rnaseq[rnaseq$HIV == "yes", 1:45]
# miseq
miseq <- read_excel( "miseq.xlsx" )
head(miseq)
# miseq[miseq$HIV == "no", 3:5]
# miseq[miseq$HIV == "yes", 3:5]
# Xs
Clin <- read_excel("clin_microbiome.xlsx")
# covariates
# vgenes
vgenes <- read_excel("/home/guanshim/Documents/gitlab/Cario_RNASeq_Microbiom_Inte/DataProcessed/Vgenes/vgenes_rld.xlsx")
head(vgenes)
dim(vgenes)
head(Clin[,1:6])
##
## clinical parameters
clinical_names <- colnames(Clin)[5:18]
clinical_names

n_clinical <- length(clinical_names)
head(Clin[18:19])
Clin$ID

```

```{r interactionlm}
########################## linear regression ##############################
## equal length of outcomes and covariates
lm_Reg_Interaction <- function(gene_matrix, Clin, clin_var_name, inter_term_name){
  ######################## If there are non-numeric columns ############### 
  gene_matrix = gene_matrix %>% select(-HIV, -ID)
  ######################
  ## Binary Interaction
  ######################
  # matrix
  gene_matrix = as.matrix(gene_matrix)
  # get names ready
  genelistname = base::colnames(gene_matrix)
  ## number of gene to test, also the number of multiple test
  n_gene = ncol(gene_matrix)
  clinical_variable = as.matrix(Clin[, colnames(Clin) == clin_var_name])
  ####### the interaction variable , here a binary variable 
  interaction_var = as.matrix(Clin[, colnames(Clin) == inter_term_name])
  ref = levels(factor(interaction_var))[1]
  non_ref = levels(factor(interaction_var))[2]
  interaction_var_ = factor(interaction_var, levels = c(non_ref, ref ))
  ## outcome lm
  outcome_lm = lapply(1:n_gene, function(i){
  lm = lm(gene_matrix[,i]~ clinical_variable*interaction_var + Clin$Age + Clin$Gender )
  ## relevel to get estimate beta1 + beta3 
  ## https://stats.stackexchange.com/questions/248248/test-whether-the-sum-of-two-coefficients-is-significantly-different-that-zero-in
  lm_relevel = lm(gene_matrix[,i]~ clinical_variable*interaction_var_ + Clin$Age + Clin$Gender )
  ##### interaction term is always the last row: 6 ########
  ####### -3: means no t-statistics, -2: means no Std.Error term #############
  coef = c(summary(lm)$coefficients[2, -c(2,3) ], 
           summary(lm_relevel)$coefficients[2, -c(2,3) ], 
           summary(lm)$coefficients[6, -c(2,3) ] )
  return(coef)
})
      outcome_lm = data.frame(matrix(unlist(outcome_lm), ncol = 6, byrow = TRUE,
                          dimnames = list(
                     c(colnames(gene_matrix)),
                        c("slope_no", "p.value_no", 
                          "slope_hiv", "p.value_hiv",
                          "slope_diff", "p.value_diff" ))))

   # adjusted p-value
    outcome_lm =  outcome_lm %>% 
      # https://cran.r-project.org/web/packages/mutoss/mutoss.pdf
      # The Benjamini-Liu (BL) step-down procedure
                              dplyr::mutate(FDR_no = if(n_gene >=  17){
                                                   p.adjust(p.value_no, "BH", n_gene )
                                                       } else if (n_gene <= 10) {p.value_no } 
                                                         else{
                                                         BL(p.value_no , alpha = 0.05, silent = T)$adjPValues
                                                         },
                                            FDR_hiv = if(n_gene >=  17){
                                                   p.adjust(p.value_hiv, "BH", n_gene )
                                                       } else if (n_gene <= 10) {p.value_hiv } 
                                                         else{
                                                         BL(p.value_hiv, alpha = 0.05, silent = T)$adjPValues
                                                         }, 
                                            FDR_diff = if(n_gene >=  17){
                                                   p.adjust(p.value_diff, "BH", n_gene )
                                                       } else if (n_gene <= 10) {p.value_diff } 
                                                         else{
                                                         BL(p.value_diff, alpha = 0.05, silent = T)$adjPValues
                                                         }, 
                                               names = colnames(gene_matrix)) %>% 
                                               dplyr::mutate(slope_no = round(slope_no, 10),
                                                      slope_hiv = round(slope_hiv, 10),
                                                      slope_diff = round(slope_diff, 10)
                                                      )%>% 
                                               mutate(sig = ifelse( (FDR_diff  < 0.05) |  (FDR_hiv  < 0.05),
                                                                    "Sig.", "Non") ) %>%
                                               select(names, everything())
   # sort by p.value
   # outcome_lm = outcome_lm[order(outcome_lm$p.value), ]
 
   ## sample size 
    size = sum(!is.na(clinical_variable))

   ## summary table 
   return(list(results = data.frame(outcome_lm), size = size, clinical = clin_var_name ))
}

###################### output summary function ####################### 
## output is the 
output_sum <- function(output, genelist){
  # genelist is the name of the genelist 
  # genelist: "RNAseq" or "MiSeq"
  # a list of two sublists
  clinical = output$clinical
  clinical = gsub("\n", "_", clinical, fixed=T)
  # find a good file name
  para_name = paste(unlist(strsplit(clinical, " ", fixed = T ))[1:3], collapse = "_")
  para_name1 = gsub("/", "_", para_name, fixed=T)
  size = output$size
  input = output$results
  # lm results
  sig = sum(input$sig == "Sig.")
  
            write.xlsx(input,
           paste("~/Documents/gitlab/Cario_RNASeq_Microbiom_Inte/DataProcessed/Vgenes_inter/", 
               sep = "", Sys.Date(),"_", para_name1, "_", genelist,"_",".xlsx"),
           sheetName= paste(para_name1,sep = "_", genelist ))
  
  if(sig > 0){


  ############# summary table #####################
  genes = input$names[input$sig == 'Sig.']
  fdr_no = format.pval(input$FDR_no[input$sig == 'Sig.'], digits=4, eps = 0.0001, scientific = F)
  fdr_hiv = format.pval(input$FDR_hiv[input$sig == 'Sig.'], digits=4, eps = 0.0001, scientific = F)
  fdr_diff = format.pval(input$FDR_diff[input$sig == 'Sig.'], digits=4, eps = 0.0001, scientific = F)
  # neg_log10p = -log10(input$p.value[input$sig == 'Sig.'])
  slope_no = format(input$slope_no[input$sig == 'Sig.'], digits = 4, scientific = T )
  slope_hiv = format(input$slope_hiv[input$sig == 'Sig.'], digits = 4, scientific = T )
  slope_diff = format(input$slope_diff[input$sig == 'Sig.'], digits = 4, scientific = T )
  sum_a = cbind(clinical, genes, fdr_no,  slope_no, 
                fdr_hiv,  slope_hiv,
                fdr_diff,  slope_diff, size)
  colnames(sum_a) = c("Parameter", "Gene", 
                      "Adjusted_P_Non", "Slope_Non", 
                      "Adjusted_P_HIV", "Slope_HIV", 
                      "Adjusted_P_Diff", "Slope_Diff", "Sample_Size")
  return(sum_a)
  }else{}

}


```


## Results

All results of RNAseq are negative.

```{r run}
############# with clinical viarable #################
############## rna seq ######
## both
clinical_sum <- matrix(NA, 0, 9)
for(i in clinical_names[1:12]) {
   output =  lm_Reg_Interaction(rnaseq, Clin, i, "HIV")
   # summary table 
   sum_a = output_sum(output, "RNAseq")
   clinical_sum = rbind(clinical_sum, sum_a)
}
kable(clinical_sum, caption = "Vgenes RNAseq")

########## Miseq ################
for(i in clinical_names[1:12]) {
   output =  lm_Reg_Interaction(miseq, Clin, i, "HIV")
   # summary table 
   sum_a = output_sum(output, "MiSeq")
   clinical_sum = rbind(clinical_sum, sum_a)
}
kable(clinical_sum, caption = "Vgenes MiSeq")

###################### with Microbiome ############
# classes of microbiome
phylum <- colnames(Clin)[1:9 + 18]
family <- colnames(Clin)[10:20 + 18]
genus <- colnames(Clin)[21:28 + 18]
species <- colnames(Clin)[29:44 + 18]

```

## Test of the function

```{r testInter, echo=FALSE, eval=FALSE}
gene_matrix <- as.matrix(rnaseq[, -c(46,47)] )
n_gene = ncol(gene_matrix)
## basic output
hiv = Clin$HIV
levels(factor(hiv))[1]
lm = lm(gene_matrix[,1] ~ Clin$`CD4 T cells (% viable)` + Clin$`CD4 T cells (% viable)`: hiv  + hiv  + Clin$Age + Clin$Gender)
summary(lm)$coefficients
## change reference level
hiv_ <- factor(hiv, levels = c("yes", "no"))
lm2 = lm(gene_matrix[,1] ~ Clin$`CD4 T cells (% viable)` + Clin$`CD4 T cells (% viable)`: hiv_  + hiv_  + Clin$Age + Clin$Gender)
summary(lm2)$coefficients



paste("The interaction term is always the last row")

outcome_lm = lapply(1:n_gene, function(i){
  lm = lm(gene_matrix[,i]~ Clin$`CD4 T cells (% viable)`*Clin$HIV + Clin$Age + Clin$Gender )
  lm2 = lm(gene_matrix[,i] ~ Clin$`CD4 T cells (% viable)` + Clin$`CD4 T cells (% viable)`: hiv_  + hiv_  + Clin$Age + Clin$Gender)
  coef = c(summary(lm)$coefficients[2, -c(2,3) ], 
           summary(lm2)$coefficients[2, -c(2,3) ], 
           summary(lm)$coefficients[6, -c(2,3) ] )
  return(coef)
})
      outcome_lm = data.frame(matrix(unlist(outcome_lm), ncol = 6, byrow = TRUE,
                          dimnames = list(
                     c(colnames(gene_matrix)),
                        c("slope_no", "p.value_no", 
                          "slope_hiv", "p.value_hiv",
                          "slope_diff", "p.value_diff" ))))
outcome_lm
   
outcome_lm  %>% dplyr::mutate(FDR = if(n_gene >=  17){
                                                   p.adjust(p.value, "BH", n_gene )
                                                       } else if (n_gene <= 10) {p.value} 
                                                         else{
                                                         BL(p.value, alpha = 0.05, silent = T)$adjPValues
                                                         },
                                            FDR_Inter = if(n_gene >=  17){
                                                   p.adjust(p.value_Inter, "BH", n_gene )
                                                       } else if (n_gene <= 10) {p.value_Inter} 
                                                         else{
                                                         BL(p.value_Inter, alpha = 0.05, silent = T)$adjPValues
                                                         }) %>% 
                                               mutate(sig = ifelse( (FDR  < 0.05) |  (FDR_Inter  < 0.05), 
                                                                     "Sig.", "Non") )

test = lm_Reg_Interaction(rnaseq, Clin, clinical_names[1], "HIV")
test$results
######## Gamma Distribution ################
###### this is only for the Outcome Single Mutation per unique Sequence #############
######### Single Mutation per unique Sequence is positive continuous data ######
glmGamma <- glm(as.matrix(rnaseq[,1]) ~ Clin$`CD4 T cells (% viable)`*Clin$HIV + Clin$Age + Clin$Gender, 
                family = Gamma(link = "log") )
summary(glmGamma)$coefficients
paste("The interaction term is always the last row")
## rnaseq
sum(rnaseq$ID != miseq$ID)
sum(rnaseq$ID != Clin$ID )
sum(rnaseq$HIV != Clin$HIV )
# as.matrix(rnaseq[,1])
# anyNA(gene_matrix)
# apply(gene_matrix, 2, anyNA )

```


