---
title: "HIV Treated vs Control"
author: "Guannan Shen"
date: "April 28, 2019"
output: 
  html_document:
    number_sections: yes
    theme: united
    toc: yes
    toc_depth: 5
    toc_float: yes
  word_document:
    toc: yes
    toc_depth: '5'
  pdf_document:
    latex_engine: lualatex
    number_sections: yes
    toc: yes
    toc_depth: 5
---

```{r setup, include=FALSE, cache = FALSE}
require("knitr")
opts_chunk$set(tidy.opts=list(width.cutoff=60),tidy=TRUE)
opts_chunk$set(engine = "R")
knitr::opts_chunk$set(echo = T)
knitr::opts_chunk$set(message = F)
knitr::opts_chunk$set(warning = F)
## setting wd in DELL
## opts_knit$set(root.dir = "~/Stats/CIDA_OMICs/CIDA_OMICS/7659Stats_Genetics/HW5/")
## setting working directory in asus 
## opts_knit$set(root.dir = "C:/Users/hithr/Documents/Stats/gitlab/Cario_RNASeq_Microbiom_Inte/DataRaw/") 
## setting working directory in ubuntu
opts_knit$set(root.dir = "~/Documents/gitlab/Cario_RNASeq_Microbiom_Inte/DataRaw/interferome/")
                                                 
## cache = F, if cache = T, will not revaluate code chunk everytime
## double or more space to insert a line break

```



```{r libs, warning=FALSE, echo=F}
######################
## Set up workspace
######################
rm(list = ls())
library(edgeR)
library(EDASeq)
library(DESeq2)
library(knitr)
library(tidyverse)
library(magrittr)
library(stats)
library(BiocParallel)
library(readxl)
library(openxlsx)
library(limma)
library(scater)
library(rgl)
library(pca3d)
library(EnhancedVolcano)
library(RColorBrewer)
options(stringsAsFactors = F)
options(dplyr.width = Inf)
getwd()
## not in function
'%nin%' <- Negate('%in%')

# ######## clean memory ######################
# rm(list = ls())
# gc()
# is(dds)
# slotNames(dds)

```

# Data Preprocessing
15 HIV-infected-treated donors vs. 11 healthy controls.  

Filtering was applied. The inclusion criteria is the gene must have a total counts (sum up all samples) more than or equal to the number of samples times 5.  


```{r eda_qc, echo=F}
# import data
cnts.raw <- read.xlsx("HIV-1_infected_HAART_treated_Raw_Counts.xlsx")
dim(cnts.raw)

########### need to realign the samples ###################

# filtering, the same 
## filter the raw data and check dim
cnts_fsym <- cnts.raw[rowSums(cnts.raw[, 4:29])>=(5*ncol(cnts.raw[, 4:29])), ]
dim(cnts_fsym)
row.names(cnts_fsym) <- NULL

### get matrix
cnts_f <- cnts_fsym %>% dplyr::select(-c(Symbol, Length ) ) %>% 
            tibble::column_to_rownames("Gene_ID") %>% as.matrix()

########### pheno grouping ####################
dim(cnts_f)
ctrl <- c("Dys273",	"Dys458",	"TR1310",	"TR1407",	"TR1411",	"TR1413",	"TR1414",	"TR1441",	"TR1465",	"TR1470",	"TR1547")
ctrl.id <-  colnames(cnts_f)[1:11]
sum(ctrl != ctrl.id)

hiv.id <- colnames(cnts_f)[12:26]
hiv <- c("TR1540",	"TR1543",	"TR1549",	"TR1551",	"TR1584",	"TR1609",	"TR1624",	"TR1667",	"TR1745",	"TR1750",	"TR1761",	"TR1762", "TR1767",	"TR1772", "TR1778" )
sum(hiv != hiv.id)

rna.pid <- colnames(cnts_f)
pheno <- data.frame(pid = rna.pid, txt = as.factor(c(rep("Control", 11), 
                                                     rep("HIV-Treated", 15)
                                                     )) )
pheno$txt %<>% relevel("Control")

############ EDA ############
# using the function from EDASeq
set <- newSeqExpressionSet(as.matrix(cnts_f),phenoData = data.frame(condition=as.factor(pheno$txt), row.names=colnames(cnts_f)))

################# deseq2 ###########################
## now using deseq2
dds <- DESeqDataSetFromMatrix(countData = counts(set), colData = pData(set),design = ~ condition)
dds <- estimateSizeFactors(dds)
## normalization factors
paste("size Factors of DESeq2")
sizeFactors(dds)
cnts.deseq2 <- counts(dds, normalized=TRUE)

######## edgeR TMM ###############
## edgeR object
group <- c(rep(1, 11), rep(2, 15))
y <- DGEList(counts= as.matrix(cnts_f), group=group)
## normalization 
y <- calcNormFactors(y, method = "TMM")
paste("Scaling Factors of edgeR")
y$samples
cnts.edger <- edgeR::cpm(y)

########## TPM #########
cnts.tpm <- calculateTPM(cnts_f, effective_length = cnts_fsym$Length)

###########3 using scater for QC EDA ##################
rna_info <- data.frame(pid = rna.pid, Group = as.factor(c(rep("Control", 11), 
                                                     rep("HIV-Treated", 15)
                                                     )) )
rna_info$Group %<>% relevel("Control")

## cnts. list
dir <- "~/Documents/gitlab/Cario_RNASeq_Microbiom_Inte/DataProcessed/Summary/Treated/plots/"

cnts.list <- list(cnts_f, cnts.edger, cnts.deseq2, cnts.tpm)
cnts.names <- c("Filtered Counts", "TMM Normalized Counts",
                "DESeq2 Normalized Counts", "TPM Normalized Counts")
par(mfrow = c(1,2))
###### OR i in 1:4 for all plots 
for (i in 2){
  j  = cnts.list[[i]]
  ## using scater, start with an object
example_sce <- SingleCellExperiment(
                 assays = list(counts = j),
                 colData = rna_info)
## RLE plots
print( scater::plotRLE(example_sce, exprs_values = "counts", exprs_logged = FALSE,
        colour_by = "Group", style = "minimal") + scale_x_discrete("Samples", labels = rna.pid) + labs(caption = paste("RLE Plot:" ,cnts.names[i]) ) )
ggsave(paste(dir, "RLE_plot_2outliers.tiff", sep = "") , dpi = 300)
## pca plots
example_sce <- SingleCellExperiment(
                 assays = list(logcounts = j),
                 colData = rna_info)
print( scater::plotPCA(example_sce, colour_by = "Group", size_by = "Group") + labs(caption = paste(  "PCA Plot:",cnts.names[i]) ))
ggsave(paste(dir, "PCA_plot_2outliers.tiff", sep = "") , dpi = 300)
}

######### the whole gene list is a fair list###########
## define the gene list 
invitro_genes <- c("Mx2", "APOBEC3G", "Siglec1", "ISG15", "Bst2")
## turn in to upper case
invitro_genes[ sapply(invitro_genes, toupper) %nin% cnts_fsym$Symbol ]

########## a bad sample ########
# using the function from EDASeq
set <- newSeqExpressionSet(as.matrix(cnts.edger),phenoData = data.frame(condition=as.factor(pheno$txt), row.names=colnames(cnts_f)))
##  general QC images  ## 
## plotRLE from EDASeq
## boxplots of the log-ratios of the gene-level read counts of each sample to those of a reference sample
# (defined as the median across the samples). 
par(mfrow = c(1,1))
EDASeq::plotRLE(set, 
        outline = FALSE, col=c(rep("steelblue", 11), rep("tomato", 15)), 
        main = "Control vs. HIV-Treated RLE Plot (TMM Normalized)", 
        xlab = "Sample", 
        ylab = "Relative Log Ratio")
## this should be deleted
rna.pid[21]

################## delete this one ####################
cnts_f_dele <- cnts_f[, -21]
dim(cnts_f_dele)
ctrl.id <-  colnames(cnts_f_dele)[1:11]
hiv.id <- colnames(cnts_f_dele)[12:25]

rna.pid <- colnames(cnts_f_dele)
pheno <- data.frame(pid = rna.pid, txt = as.factor(c(rep("Control", 11), 
                                                     rep("HIV-Treated", 14)
                                                     )) )
pheno$txt %<>% relevel("Control")

# using the function from EDASeq
set <- newSeqExpressionSet(as.matrix(cnts_f_dele),phenoData = data.frame(condition=as.factor(pheno$txt), row.names=colnames(cnts_f_dele)))

################# deseq2 ###########################
## now using deseq2
dds <- DESeqDataSetFromMatrix(countData = counts(set), colData = pData(set),design = ~ condition)
dds <- estimateSizeFactors(dds)
## normalization factors
paste("size Factors of DESeq2")
sizeFactors(dds)
cnts.deseq2 <- counts(dds, normalized=TRUE)

######## edgeR TMM ###############
## edgeR object
group <- c(rep(1, 11), rep(2, 14))
y <- DGEList(counts= as.matrix(cnts_f_dele), group=group)
## normalization 
y <- calcNormFactors(y, method = "TMM")
paste("Scaling Factors of edgeR")
y$samples
cnts.edger <- edgeR::cpm(y)

########## TPM #########
cnts.tpm <- calculateTPM(cnts_f_dele, effective_length = cnts_fsym$Length)

###########3 using scater for QC EDA ##################
rna_info <- data.frame(pid = rna.pid, Group = as.factor(c(rep("Control", 11), 
                                                     rep("HIV-Treated", 14)
                                                     )) )
rna_info$Group %<>% relevel("Control")

## cnts. list
cnts.list <- list(cnts_f_dele, cnts.edger, cnts.deseq2, cnts.tpm)
cnts.names <- c("Filtered Counts", "TMM Normalized Counts",
                "DESeq2 Normalized Counts", "TPM Normalized Counts")
par(mfrow = c(1,2))
for (i in 2){
  j  = cnts.list[[i]]
  ## using scater, start with an object
example_sce <- SingleCellExperiment(
                 assays = list(counts = j),
                 colData = rna_info)
## RLE plots
print( scater::plotRLE(example_sce, exprs_values = "counts", exprs_logged = FALSE,
        colour_by = "Group", style = "minimal") + scale_x_discrete("Samples", labels = rna.pid) + labs(caption = paste("RLE Plot:" ,cnts.names[i]) ) )
ggsave(paste(dir, "RLE_plot_outlier.tiff", sep = "") , dpi = 300)
## pca plots
example_sce <- SingleCellExperiment(
                 assays = list(logcounts = j),
                 colData = rna_info)
print( scater::plotPCA(example_sce, colour_by = "Group", size_by = "Group") + labs(caption = paste(  "PCA Plot:",cnts.names[i]) ))
ggsave(paste(dir, "PCA_plot_outlier.tiff", sep = "") , dpi = 300)
}

```

From the scaling factor, we can tell the between-samples variance is samller compared with the HIV-untreated dataset. For the previous dataset, the more the size factors differ, the more residual dependence of the variance on the mean will be found in the transformed data. **rlog** is a transformation which can perform better in these cases.  

For this dataset, a variance stabilizing transformation, **vst** should be more proper.  

varianceStabilizingTransformation(object, blind = FALSE, fitType = "parametric")  

getVarianceStabilizedData(object)

From the quality control plots, "TR1750" was deleted. 

```{r de_edger}
# make sure y is the object from edger
## DE analysis of edger
y <- estimateDisp(y)
et <- exactTest(y, pair = 1:2)
dim(et)
y$samples
### results
topTags(et, n = 10, adjust.method = "BH")
head(et$table)
et$comparison
## summary results
summary(decideTests(et, p.value = 0.05,
            lfc = 0))
# paste(1706+ 2540, "genes with FDR 0.05")

summary(decideTests(et, p.value = 0.05,
            lfc = 1))
# paste(157 + 680, "genes with FDR 0.05")

summary(decideTests(et, p.value = 0.05,
            lfc = 2))
########## deleted ###########
"TR1750"
## import genes lists 
isgs <- as.data.frame(read.delim("~/Documents/gitlab/Cario_RNASeq_Microbiom_Inte/DataRaw/coreISG252"))
dim(isgs)
genesbeta <- as.data.frame(read.delim("~/Documents/gitlab/Cario_RNASeq_Microbiom_Inte/DataRaw/genesbeta424"))
dim(genesbeta)
sig_res <- as.data.frame(topTags(et, n = 30, adjust.method = "BH")[[1]])

write.xlsx(sig_res,
           "~/Documents/gitlab/Cario_RNASeq_Microbiom_Inte/DataProcessed/interferome/treated/DE_results_treated.xlsx", sheetName="DE Results Treated")


sig_res_format <- sig_res %>% data.frame() %>% rownames_to_column("Gene_ID")

merge(sig_res_format,  isgs, by = "Gene_ID")
merge(sig_res_format,  genesbeta, by = "Gene_ID")

merge(sig_res_format, cnts_fsym[, c(1,2)], by = "Gene_ID") %>%
            dplyr::select(-Gene_ID) %>% dplyr::select( Symbol, everything())  %>%
                dplyr::arrange(PValue)  %>%   dplyr::mutate(Direction = ifelse(logFC > 0,
                                                                               "Up in HIV",
                                                                               "Down in HIV") ) %>%
                      kable

```

The DE analysis, only 30 out of 19454 genes are significantly altered between hiv-status.  
In terms of Healthy Controls vs. HIV-Infected:  
Comparison of groups:  2-1   

[1] "1" "2"

         2-1
         
Down    1706

NotSig 15644

Up      2540

[1] "4246 genes with FDR 0.05"

         2-1
         
Down     157

NotSig 19053

Up       680

[1] "837 genes with FDR 0.05 and LogFC 1"


```{r scatter}
# all the results 
res.edger <- et$table %>% dplyr::mutate(Symbol = cnts_fsym$Symbol,
                                 FDR = p.adjust(et$table$PValue, method = "BH") )
# label 
group.plot <- factor(c(rep(1,11), rep(2,14)), labels = c("Health Control", "HIV-Treated"))
## use the # No. of genes in the cnts.edger matrix
sim_plotcounts <- function(genenumber){
  plot(as.integer(group.plot)+ runif(ncol(cnts.edger), -0.1, 0.1), cnts.edger[genenumber, ], 
    xlim = c(0.5, length(levels(group.plot)) + 0.8), xaxt="n" ,
    xlab = "Groups", ylab = "Normalized Counts", main = res.edger[genenumber, ]$Symbol)
axis(1, at = seq_along(levels(group.plot)), levels(group.plot))
## pos below
text(length(levels(group.plot)) + 0.5, stats::quantile( cnts.edger[genenumber, ], 0.95), 
     paste("log2 Fold Change:", round(res.edger[genenumber, ]$logFC,2),
             "","Adjusted p-value:",  
           format(res.edger[genenumber, ]$FDR,digits=2,scientific=TRUE ),"",
           "Gene_ID: ", rownames(cnts.edger)[genenumber],
                                           sep = "\n"), pos = 1)
}
## plot top 30
for(i in order(res.edger$PValue)[1:30]){
  sim_plotcounts(i)
}

## check the outlier
outliers <- c("SNORD3A", "SNORA12", "MLLT6", "NMT2",
                                          "MAN2C1", "EDRF1-AS1", "CTD-2373H9.3",
                                          "RP3-522J7.7", "RP11-848P1.4", "LINC00854")
# get the gennumber of outliers
out_number <- res.edger[order(res.edger$PValue)[1:30], ] %>% rownames_to_column("Num") %>% 
   dplyr::filter(Symbol %in% outliers)
# get cnts especially for healthy controls
cnts_out <- cnts.edger[as.numeric(out_number$Num), 1:11] %>% as.data.frame() 

#
out_id <- NULL
for(i in 1: nrow(cnts_out)){
  out_id <- c(out_id, colnames(cnts_out)[cnts_out[i, ] == max(cnts_out[i,])])
}
length(out_id)
out_id


# 
kable(cnts_out[, 1:7])
paste("Basically, TR1310 is the outlier based on the grouped scattered plot")

```

# Final Analysis
According to the quality control plots, "TR1750" in HIV-treated was deleted.  
According to the PCA plots and grouped scatter plots of top DE genes, "TR1310" in healthy controls should be deleted. 


```{r final}
# this one should be further deleted based on 
# cnts_f_dele
which( rna.pid %in% "TR1310" )

########## need to double check by the pca plots scater::plotPCA ################## 
paste("Before Deletion")
for (i in 2){
  j  = cnts.list[[i]]
  ## using scater, start with an object
example_sce <- SingleCellExperiment(
                 assays = list(counts = j),
                 colData = rna_info)
## RLE plots
print( scater::plotRLE(example_sce, exprs_values = "counts", exprs_logged = FALSE,
        colour_by = "Group", style = "minimal") + scale_x_discrete("Samples", labels = rna.pid) + labs(caption = paste("RLE Plot:" ,cnts.names[i]) ) )
## pca plots
example_sce <- SingleCellExperiment(
                 assays = list(logcounts = j),
                 colData = rna_info)
print( scater::plotPCA(example_sce, colour_by = "Group", size_by = "Group") + labs(caption = paste(  "PCA Plot:",cnts.names[i]) ))
}


################## delete this one ####################
cnts_f_f <- cnts_f_dele[, -3]
dim(cnts_f_f)
ctrl.id <-  colnames(cnts_f_f)[1:10]
hiv.id <- colnames(cnts_f_f)[11:24]

rna.pid <- colnames(cnts_f_f)
pheno <- data.frame(pid = rna.pid, txt = as.factor(c(rep("Control", 10), 
                                                     rep("HIV-Treated", 14)
                                                     )) )
pheno$txt %<>% relevel("Control")

# using the function from EDASeq
set <- newSeqExpressionSet(as.matrix(cnts_f_f),phenoData = data.frame(condition=as.factor(pheno$txt), row.names=colnames(cnts_f_f)))

# 
######## edgeR TMM ###############
## edgeR object
group <- c(rep(1, 10), rep(2, 14))
y <- DGEList(counts= as.matrix(cnts_f_f), group=group)
## normalization 
y <- calcNormFactors(y, method = "TMM")
paste("Scaling Factors of edgeR")
y$samples
# the TMM normalized Counts
cnts.edger <- edgeR::cpm(y)
cnts.TMM.id <- cnts.edger %>% as.data.frame() %>% rownames_to_column("Gene_ID")
sum(cnts.TMM.id$Gene_ID == cnts_fsym$Gene_ID)


# save TMM normalized counts 
write.xlsx(merge(cnts.TMM.id, cnts_fsym[, 1:2], by = "Gene_ID"),
           "~/Documents/gitlab/Cario_RNASeq_Microbiom_Inte/DataProcessed/Summary/Treated/TMM_counts_HIVtreated.xlsx",
           sheetName="TMM HIV treated")

## cnts. list
# all remains the sames other than cnts.edger
# since we will only use edger for downstream analysis
cnts.list <- list(cnts_f_dele, cnts.edger, cnts.deseq2, cnts.tpm)
cnts.names <- c("Filtered Counts", "TMM Normalized Counts",
                "DESeq2 Normalized Counts", "TPM Normalized Counts")
par(mfrow = c(1,2))

rna_info <- data.frame(pid = rna.pid, Group = as.factor(c(rep("Control", 10), 
                                                     rep("HIV-Treated", 14)
                                                     )) )
rna_info$Group %<>% relevel("Control")
dir <- "~/Documents/gitlab/Cario_RNASeq_Microbiom_Inte/DataProcessed/Summary/Treated/plots/"
# qc plots
for (i in 2){
  j  = cnts.list[[i]]
  ## using scater, start with an object
example_sce <- SingleCellExperiment(
                 assays = list(counts = j),
                 colData = rna_info)
## RLE plots
print( scater::plotRLE(example_sce, exprs_values = "counts", exprs_logged = FALSE,
        colour_by = "Group", style = "minimal") + scale_x_discrete("Samples", labels = rna.pid) + labs(caption = paste("RLE Plot:" ,cnts.names[i]) ) )
ggsave(paste(dir, "RLE_plot.tiff", sep = "") , dpi = 300)
## pca plots
example_sce <- SingleCellExperiment(
                 assays = list(logcounts = j),
                 colData = rna_info)
print( scater::plotPCA(example_sce, colour_by = "Group", size_by = "Group") + labs(caption = paste(  "PCA Plot:",cnts.names[i]) ))
ggsave(paste(dir, "PCA_plot.tiff", sep = "") , dpi = 300)
}


############### DE analysis ########################
## DE analysis of edger
y <- estimateDisp(y)
et <- exactTest(y, pair = 1:2)
dim(et)
y$samples
### results
topTags(et, n = 10, adjust.method = "BH")
head(et$table)
et$comparison
## summary results
summary(decideTests(et, p.value = 0.2,
            lfc = 0))

summary(decideTests(et, p.value = 0.1,
            lfc = 0))

summary(decideTests(et, p.value = 0.05,
            lfc = 0))
# paste(1706+ 2540, "genes with FDR 0.05")

summary(decideTests(et, p.value = 0.05,
            lfc = 1))
# paste(157 + 680, "genes with FDR 0.05")

summary(decideTests(et, p.value = 0.05,
            lfc = 2))

################# results #############
## import genes lists 
isgs <- as.data.frame(read.delim("~/Documents/gitlab/Cario_RNASeq_Microbiom_Inte/DataRaw/coreISG252"))
dim(isgs)
genesbeta <- as.data.frame(read.delim("~/Documents/gitlab/Cario_RNASeq_Microbiom_Inte/DataRaw/genesbeta424"))
dim(genesbeta)
sig_res <- as.data.frame(topTags(et, n = 30, adjust.method = "BH")[[1]])

sig_res_format <- sig_res %>% data.frame() %>% rownames_to_column("Gene_ID")

write.xlsx(sig_res_format,
           "~/Documents/gitlab/Cario_RNASeq_Microbiom_Inte/DataProcessed/Summary/Treated/DE_results_treated_005.xlsx", sheetName="DE Results Treated")

# relax the cutoff to 0.05
merge(sig_res_format,  isgs, by = "Gene_ID")
merge(sig_res_format,  genesbeta, by = "Gene_ID")

# relax the cutoff to 0.2
sig_res <- as.data.frame(topTags(et, n = 121, adjust.method = "BH")[[1]])

sig_res_format <- sig_res %>% data.frame() %>% rownames_to_column("Gene_ID")

results.isgs <- merge(sig_res_format,  isgs, by = "Gene_ID")
results.isgs
write.xlsx(results.isgs,
           "~/Documents/gitlab/Cario_RNASeq_Microbiom_Inte/DataProcessed/Summary/Treated/DE_isgs_results_treated_02.xlsx", sheetName="DE Results ISGS Treated")
results.genesbeta <-  merge(sig_res_format,  genesbeta, by = "Gene_ID")
results.genesbeta
write.xlsx(results.genesbeta,
           "~/Documents/gitlab/Cario_RNASeq_Microbiom_Inte/DataProcessed/Summary/Treated/DE_genesbeta_results_treated_02.xlsx", sheetName="DE Results Beta Treated")

results <- merge(sig_res_format, cnts_fsym[, c(1,2)], by = "Gene_ID") %>% 
            dplyr::select(-Gene_ID) %>% dplyr::select( Symbol, everything())  %>% 
                dplyr::arrange(PValue)  %>%   dplyr::mutate(Direction = ifelse(logFC > 0, 
                                                                               "Up in HIV",
                                                                               "Down in HIV") ) 

results %>% kable

write.xlsx(results,
           "~/Documents/gitlab/Cario_RNASeq_Microbiom_Inte/DataProcessed/Summary/Treated/DE_results_treated_02.xlsx", sheetName="DE Results Treated")

####### save all #########
# all results
all_res<- as.data.frame(topTags(et, n = nrow(cnts_fsym), adjust.method = "BH")[[1]]) %>% 
      data.frame() %>% rownames_to_column("Gene_ID")
all_res_format <- merge(cnts_fsym[, 1:2], all_res, by = "Gene_ID")
all_res_format %>% data.frame() %>%
            dplyr::select(-Gene_ID) %>% dplyr::select( Symbol, everything())  %>%
                dplyr::arrange(PValue)  %>%   dplyr::mutate(Direction = ifelse(logFC > 0,
                                                                               "Up in HIV",
                                                                               "Down in HIV") ) %>%
  write.xlsx(.,
           "~/Documents/gitlab/Cario_RNASeq_Microbiom_Inte/DataProcessed/Summary/Treated/DE_results_treated_all.xlsx", sheetName="DE Results Treated")
                

############### grouped scatter plot ###################
par(mfrow = c(1,1))
# all the results 
res.edger <- et$table %>% dplyr::mutate(Symbol = cnts_fsym$Symbol,
                                 FDR = p.adjust(et$table$PValue, method = "BH") )
# label 
group.plot <- factor(c(rep(1,10), rep(2,14)), labels = c("Health Control", "HIV-Treated"))
for(i in order(res.edger$PValue)[1:30]){
  sim_plotcounts(i)
  # tiff( paste ("~/Documents/gitlab/Cario_RNASeq_Microbiom_Inte/DataProcessed/Summary/Treated/plots/grouped_scattter",i, ".tiff", sep = "") )
  # dev.off()
}



```