---
title: "Untreated Plots Tables"
author: "Guannan Shen"
date: "April 29, 2019"
output: 
  html_document:
    number_sections: yes
    theme: united
    toc: yes
    toc_depth: 5
    toc_float: yes
  word_document:
    toc: yes
    toc_depth: '5'
  pdf_document:
    latex_engine: lualatex
    number_sections: yes
    toc: yes
    toc_depth: 5
---

```{r setup, include=FALSE, cache = FALSE}
require("knitr")
opts_chunk$set(tidy.opts=list(width.cutoff=60),tidy=TRUE)
opts_chunk$set(engine = "R")
knitr::opts_chunk$set(echo = T)
knitr::opts_chunk$set(message = F)
knitr::opts_chunk$set(warning = F)
## setting wd in DELL
## opts_knit$set(root.dir = "~/Stats/CIDA_OMICs/CIDA_OMICS/7659Stats_Genetics/HW5/")
## setting working directory in asus 
## opts_knit$set(root.dir = "C:/Users/hithr/Documents/Stats/gitlab/Cario_RNASeq_Microbiom_Inte/DataRaw/") 
## setting working directory in ubuntu
opts_knit$set(root.dir = "~/Documents/gitlab/Cario_RNASeq_Microbiom_Inte/DataProcessed/interferome/")
                                                 
## cache = F, if cache = T, will not revaluate code chunk everytime
## double or more space to insert a line break
```



```{r libs, warning=FALSE, echo=F}
######################
## Set up workspace
######################
rm(list = ls())
library(edgeR)
library(EDASeq)
library(DESeq2)
library(knitr)
library(tidyverse)
library(magrittr)
library(stats)
library(BiocParallel)
library(readxl)
library(openxlsx)
library(limma)
library(scater)
library(rgl)
library(pca3d)
library(EnhancedVolcano)
library(RColorBrewer)
library(grid)
library(ggrepel)
options(stringsAsFactors = F)
options(dplyr.width = Inf)
getwd()
## not in function
'%nin%' <- Negate('%in%')

# ######## clean memory ######################
# rm(list = ls())
# gc()
# is(dds)
# slotNames(dds)

```

## Import Data
rlog transformed counts. 

```{r pheno, echo=F}
# import unnormalized counts table
cnts.raw <- read.delim("/home/guanshim/Documents/gitlab/Cario_RNASeq_Microbiom_Inte/DataRaw/All_Sample_geneCounts_raw_counts.txt", header = TRUE, sep = "\t")

cnts <- cnts.raw %>% dplyr::select(-c(Symbol, Length ) ) %>% tibble::column_to_rownames("Gene_ID")
cnts <- as.matrix(cnts)
rna.pid <- colnames(cnts)
# now we have the common counts table 
## pheno
ctrl.id <-  colnames(cnts)[1:13]
hiv.id <- colnames(cnts)[14:32]
## from dim() we know there are 32 samples
pheno <- data.frame(pid = rna.pid, txt = as.factor(c(rep("Control", 13), 
                                                     rep("HIV", 19)
                                                     )) )
pheno$txt %<>% relevel("Control")
```

```{r data}
# rlg
isgs.05.lin <- read.csv("~/Documents/gitlab/Cario_RNASeq_Microbiom_Inte/DataProcessed/corr/isgs.05.lin.csv", row.names = 1)
dim(isgs.05.lin)
genesbeta.05.lin <- read.csv("~/Documents/gitlab/Cario_RNASeq_Microbiom_Inte/DataProcessed/corr/genesbeta.05.lin.csv", row.names = 1)
dim(genesbeta.05.lin)

# clinical data
clinical_order <- read.csv("clinical_order.csv")
clinical_names <- c("Blood CD4 T Cell Counts (cells/ul)", "Plasma Viral Load", 
                    "Tissue HIV RNA (per CD4 T cell)",
                    "Tissue CD4 T Cell Counts (number/g)", "IL-6 (pg/ml)", "CRP (ug/ml)", "iFABP (pg/ml)",
                    "sCD27 (U/ml)", "CD14 (ng/ml)", "LPS (pg/ml)", "LTA (OD)", 
                    base::paste("IFN", '\u03b1', sep = "" ),  base::paste("IFN", '\u03b2', sep = "" ),
                    "CD4 T cells (% viable CD45+ cells)")

n_clinical <- length(clinical_names)
clinical_names
# previous results

```

## Test correlations and Summary

```{r corr, echo=F}
gene_IFNReg <- function(gene_matrix, clinical_variable, clin_var_name){
  # get names ready
  genelistname = base::colnames(gene_matrix)
  ## number of gene to test, also the number of multiple test
  n_gene = ncol(gene_matrix)
  ## outcome lm
  outcome_lm = lapply(1:n_gene, function(i){
    lm = lm(gene_matrix[,i]~ clinical_variable + clinical_order$age + clinical_order$sex )
    coef = summary(lm)$coefficients[2, ]
    return(coef)
  })
  outcome_lm = data.frame(matrix(unlist(outcome_lm), ncol = 4, byrow = TRUE,
                                 dimnames = list(
                                   c(colnames(gene_matrix)),
                                   c("Estimate", "Std.Error", "t.statistic", "p.value"))))
  
  # adjusted p-value
  outcome_lm =  outcome_lm %>% 
     dplyr::mutate(FDR = p.adjust(p.value, "BH", n_gene ),
                  names = colnames(gene_matrix)) %>% 
     dplyr::mutate(Estimate = round(Estimate, 10),
                  Std.Error = round(Std.Error, 10),
                  t.statistic = round(t.statistic,4))%>% 
     select(names, everything())
  # sort by p.value
  outcome_lm = outcome_lm[order(outcome_lm$p.value), ]
  
  ## sample size 
  size = sum(!is.na(clinical_variable))
  
  ## summary table 
  return(list(results = data.frame(outcome_lm), size = size, clinical = clin_var_name ))
}



```


```{r sum}
########### summary ###############
clinical_sum <- matrix(NA, 14, 14)
alpha <-  0.05
for(i in 1:n_clinical) {
  ## number of clinical virable
  j = c(6:19)[i]
  cliname = base::colnames(clinical_order)[j]
  ## linear regression
  lin_res_isgs = gene_IFNReg(isgs.05.lin, clinical_order[,j], clinical_names[i])
  lin_res_genesbeta = gene_IFNReg(genesbeta.05.lin, clinical_order[,j], clinical_names[i])
  ## save data
  
  write.xlsx(lin_res_isgs$results,
             paste("~/Documents/gitlab/Cario_RNASeq_Microbiom_Inte/DataProcessed/interferome/sum/isgs_", 
                   cliname,".xlsx", sep = ""), sheetName= paste("ISGs_", cliname, sep = "") )
  
  write.xlsx(lin_res_genesbeta$results,
             paste("~/Documents/gitlab/Cario_RNASeq_Microbiom_Inte/DataProcessed/interferome/sum/ifnbeta_", 
                   cliname,".xlsx", sep = ""), sheetName= paste("ifnbeta_", cliname, sep = ""))
  # }
  ## check basic
  if(lin_res_isgs$size == lin_res_genesbeta$size){
    print("Good size")
    size = lin_res_isgs$size
  }else stop("sample size wrong")
  if(lin_res_isgs$clinical == lin_res_genesbeta$clinical){
    print("Good outcome")
    clinical = lin_res_isgs$clinical
  }else stop("clinical parameter wrong")
  ## number of genes
  nisgs = nrow(lin_res_isgs$results)
  ngenesbeta = nrow(lin_res_genesbeta$results)
  ## results 
  isgs_lm = data.frame(lin_res_isgs$results)
  genesbeta_lm = data.frame(lin_res_genesbeta$results)
  ## xlim
  xrange = c(range(isgs_lm$Estimate)*1.3, range(genesbeta_lm$Estimate)*1.3)
  ################## summary table ###################
  ###  sig prop
  isgs.sig.no =  sum(isgs_lm$FDR < 0.05)
  genesbeta.sig.no =  sum(genesbeta_lm$FDR < 0.05)
  if( (isgs.sig.no >= 5) & (genesbeta.sig.no >= 5) ){
    prop.test.sig = prop.test(x = c(isgs.sig.no, genesbeta.sig.no), n = c(nisgs, ngenesbeta), correct = FALSE)
    sig.p = format(prop.test.sig$p.value, digits = 2, scientific = T)
  }else{
    prop.test.sig = prop.test(x = c(isgs.sig.no, genesbeta.sig.no), n = c(nisgs, ngenesbeta), correct = TRUE)
    sig.p = format(prop.test.sig$p.value, digits = 2, scientific = T)
  }
  ########### volcano plots ################
  if(min(isgs_lm$FDR) < 0.05){
    rownames(isgs_lm) <- isgs_lm$names
    p_cutoff <- isgs_lm$p.value[min(which(isgs_lm$FDR >= 0.05))]
    p1 <- ggplot(isgs_lm, aes(Estimate, -log10(p.value)) ) +
          geom_point(colour = "gray65") + 
          geom_point(data = subset(isgs_lm, FDR < 5e-2 & Estimate > 0 ),  
             aes(Estimate, -log10(p.value)), colour="black") +
          geom_point(data = subset(isgs_lm, FDR < 5e-2 & Estimate < 0 ),  
             aes(Estimate, -log10(p.value)), colour="gray35") +
          geom_hline(yintercept=-log10(p_cutoff), 
                     linetype="dashed", color = "black") +
          geom_vline(xintercept=0, linetype="dashed", color = "black") +
          labs(x = expression(paste("Slope" )),
               y =  expression(paste("-",log[10]," p-value" )),
               caption =  paste("Core ISGs: associations with ", clinical , sep = "") ) +
          annotate(geom="text", x = max(xrange)/1.2 ,  
                    y=-log10( p_cutoff - 0.005), 
                    label="FDR < 0.05",
                    color="black") +
          xlim( min(xrange) , max(xrange) ) + 
          theme_bw()
       
    ggsave(paste("Volcanol_Plot_ISGs_", cliname, sep = ""), width = 6, height = 9, device = "tiff", path = "~/Documents/gitlab/Cario_RNASeq_Microbiom_Inte/DataProcessed/interferome/plots", dpi = 300)
    print(p1)
  } else{print("No Sig.")}
  
  if(min(genesbeta_lm$FDR) < 0.05){
    rownames(genesbeta_lm) <- genesbeta_lm$names
    p_cutoff2 <- genesbeta_lm$p.value[min(which(genesbeta_lm$FDR >= 0.05))]
    p2 <- ggplot(genesbeta_lm, aes(Estimate, -log10(p.value)) ) +
          geom_point(colour = "gray65") + 
          geom_point(data = subset(genesbeta_lm, FDR < 5e-2 & Estimate > 0 ),  
             aes(Estimate, -log10(p.value)), colour="black") +
          geom_point(data = subset(genesbeta_lm, FDR < 5e-2 & Estimate < 0 ),  
             aes(Estimate, -log10(p.value)), colour="gray35") +
          geom_hline(yintercept=-log10(p_cutoff2), 
                     linetype="dashed", color = "black") +
          geom_vline(xintercept=0, linetype="dashed", color = "black") +
          labs(x = expression(paste("Slope" )),
               y =  expression(paste("-",log[10]," p-value" )),
               caption =  paste("IFN-Beta Genes: associations with ", clinical , sep = "") ) +
          annotate(geom="text", x = max(xrange)/1.2 ,  
                   y=-log10( p_cutoff2 - 0.005 ), 
                    label="FDR < 0.05",
                    color="black") +
          xlim( min(xrange) , max(xrange) ) + 
          theme_bw()
    
    ggsave(paste("Volcanol_Plot_IFNBeta_", cliname, sep = ""), width = 6, height = 9, device = "tiff", path = "~/Documents/gitlab/Cario_RNASeq_Microbiom_Inte/DataProcessed/interferome/plots", dpi = 300)
    print(p2)
  } else{print("No Sig.")}
  
  ## positive prop
  isgs.pos.no = sum(isgs_lm$Estimate > 0 & isgs_lm$FDR < 0.05)
  genesbeta.pos.no = sum(genesbeta_lm$Estimate > 0 & genesbeta_lm$FDR < 0.05)
    if(isgs.sig.no == 0 | genesbeta.sig.no == 0 ) {
       pos.p = NA
    }else{
     if( (isgs.pos.no >= 5) & (genesbeta.pos.no >= 5) ){
       prop.test.pos = prop.test(x = c(isgs.pos.no, genesbeta.pos.no), 
                              n = c(isgs.sig.no, genesbeta.sig.no), correct = FALSE)
       pos.p = format(prop.test.pos$p.value, digits = 2, scientific = T)
      }else{
       prop.test.pos = prop.test(x = c(isgs.pos.no, genesbeta.pos.no), 
                              n = c(isgs.sig.no, genesbeta.sig.no), correct = TRUE)
       pos.p = format(prop.test.pos$p.value, digits = 2, scientific = T)
      }
    }
  ## If you check prop.test, chisq.test and z-test on your data then they all give you the same p-value
  ## continuity correction if anyone less than 5
  ## the clinical summary table 
  print(clinical)
  clinical_sum[i, ] <- c(clinical, size, nisgs, ngenesbeta, isgs.sig.no,
                         round(isgs.sig.no/nisgs, 2) , genesbeta.sig.no, 
                         round(genesbeta.sig.no/ngenesbeta,2) , sig.p, isgs.pos.no,
                         round(isgs.pos.no/isgs.sig.no,2),  genesbeta.pos.no,
                         round(genesbeta.pos.no/genesbeta.sig.no,2) , 
                         pos.p)
  
  colnames(clinical_sum) <- c("Clinical Parameter", "Sample Size", "ISGs", "IFN-Beta", "No. Sig. ISGs",
                              "ISGs Sig.", "No. Sig. IFN-Betas",
                              "IFNbeta Sig.", "p value", "No. Pos. ISGs", 
                              "ISGs Pos. Prop.", "No. Pos. IFN-Betas",
                              "IFNbeta Pos. Prop.", "p value (Pos)")
}


## clinical_sum
write.xlsx(clinical_sum, 
           "~/Documents/gitlab/Cario_RNASeq_Microbiom_Inte/DataProcessed/interferome/clinical_summary.xlsx",
           sheetName = paste("FDR", alpha) )

clinical_sum[12:13,1] <- c("IFNalpha", "IFNbeta")
kable(data.frame(clinical_sum ) )


```

```{r il-6tables}
genesbeta_il6 <- read.xlsx("~/Documents/gitlab/Cario_RNASeq_Microbiom_Inte/DataProcessed/interferome/sum/ifnbeta_IL.6.xlsx")
kable(genesbeta_il6)
################### il-6 pllots ##########################
alpha <-  0.05
i = 5 # for il-6
  ## number of clinical virable
  j = c(6:19)[i]
  cliname = base::colnames(clinical_order)[j]
  
  ## linear regression
  lin_res_isgs = gene_IFNReg(isgs.05.lin, clinical_order[,j], clinical_names[i])
  lin_res_genesbeta = gene_IFNReg(genesbeta.05.lin, clinical_order[,j], clinical_names[i])
  clinical = lin_res_isgs$clinical
    ## number of genes
  nisgs = nrow(lin_res_isgs$results)
  ngenesbeta = nrow(lin_res_genesbeta$results)
  ## results 
  isgs_lm = data.frame(lin_res_isgs$results)
  genesbeta_lm = data.frame(lin_res_genesbeta$results)
  ## xlim
  xrange = c(range(isgs_lm$Estimate)*1.3, range(genesbeta_lm$Estimate)*1.3)
  ################## summary table ###################
  ###  sig prop
  isgs.sig.no =  sum(isgs_lm$FDR < alpha)
  genesbeta.sig.no =  sum(genesbeta_lm$FDR < alpha)
  ######### isgs IL-6 plots ##############
  rownames(isgs_lm) <- isgs_lm$names
    p_cutoff <- isgs_lm$p.value[min(which(isgs_lm$FDR >= alpha))]
    p1 <- ggplot(isgs_lm, aes(Estimate, -log10(p.value)) ) +
          geom_point(colour = "gray65") + 
          geom_point(data = subset(isgs_lm, FDR < alpha & Estimate > 0 ),  
             aes(Estimate, -log10(p.value)), colour="black") +
          geom_point(data = subset(isgs_lm, FDR < alpha & Estimate < 0 ),  
             aes(Estimate, -log10(p.value)), colour="gray35") +
          geom_hline(yintercept=-log10(p_cutoff), 
                     linetype="dashed", color = "black") +
          geom_vline(xintercept=0, linetype="dashed", color = "black") +
          labs(x = expression(paste("Slope" )),
               y =  expression(paste("-",log[10]," p-value" )),
               caption =  paste("Core ISGs: associations with ", clinical , sep = "") ) +
          annotate(geom="text", x = max(xrange)/1.2 ,  
                    y=-log10( p_cutoff + 0.0005), 
                    label="FDR < 0.05",
                    color="black") +
          xlim( min(xrange) , max(xrange) ) + 
          theme_bw()
    print(p1)
    ggsave(paste("Volcanol_Plot_ISGs_", cliname, sep = ""), width = 6, height = 9, device = "tiff", path = "~/Documents/gitlab/Cario_RNASeq_Microbiom_Inte/DataProcessed/interferome/plots", dpi = 300)
```

### Double Check with Scatter Plot Clinical Parameter vs. Genes 

```{r vs_clin, echo=F, eval=F}
genesbeta_il6 <- c("SCYL2", "C16orf91", "VIMP", "PTPRN2", "NUP54", "GTF2F2", "EBPL", "PTS")
# genesbeta.05.lin
# clinical_order
gene_clin_scatter_beta <- function(gene_symbol, clin_symbol){
  x = 
  p1 = ggplot(mapping = aes(x = clinical_order$IL.6 , y = pc_sample$PC1) ) +
         labs(x = "IL-6 (pg/ml)",
              y = paste("PC1 (", round(pc.sd$Variance[1],2) ,"%", " Variance", ")", sep = "") , 
         caption = paste("IL-6 (pg/ml) vs PC1 for ", name ) )  +
         geom_point(aes(shape = pheno$txt, colour = pheno$txt), size = 2.5) +
        geom_smooth(method = "loess", se = FALSE, colour = "black") +
        geom_smooth(method = "lm", se = FALSE, colour = "black") +
        scale_color_manual(values=c("gray70", "Black"), 
                    name="Groups",
                    labels = group_labels
                    ) +
       scale_shape_manual( values = c(16, 17), 
                    name="Groups",
                    labels = group_labels
                    ) +
        theme_bw()
    print(p1)
}

gene_clin_scatter_beta <- function(gene_symbol, clin_symbol){
  p1 = ggplot(mapping = aes())
    print(p1)
}


gene_clin_scatter_beta("SCYL2", "IL.6")

```

### FDR 0.1 
```{r 01sum}
########### summary ###############
clinical_sum <- matrix(NA, 14, 14)
alpha <-  0.1 
for(i in 1:n_clinical) {
  ## number of clinical virable
  j = c(6:19)[i]
  cliname = base::colnames(clinical_order)[j]
  ## linear regression
  lin_res_isgs = gene_IFNReg(isgs.05.lin, clinical_order[,j], clinical_names[i])
  lin_res_genesbeta = gene_IFNReg(genesbeta.05.lin, clinical_order[,j], clinical_names[i])
  ## save data
  
  write.xlsx(lin_res_isgs$results,
             paste("~/Documents/gitlab/Cario_RNASeq_Microbiom_Inte/DataProcessed/interferome/sum/isgs_", 
                   cliname,".xlsx", sep = ""), sheetName= paste("ISGs_", cliname, sep = "") )
  
  write.xlsx(lin_res_genesbeta$results,
             paste("~/Documents/gitlab/Cario_RNASeq_Microbiom_Inte/DataProcessed/interferome/sum/ifnbeta_", 
                   cliname,".xlsx", sep = ""), sheetName= paste("ifnbeta_", cliname, sep = ""))
  # }
  ## check basic
  if(lin_res_isgs$size == lin_res_genesbeta$size){
    print("Good size")
    size = lin_res_isgs$size
  }else stop("sample size wrong")
  if(lin_res_isgs$clinical == lin_res_genesbeta$clinical){
    print("Good outcome")
    clinical = lin_res_isgs$clinical
  }else stop("clinical parameter wrong")
  ## number of genes
  nisgs = nrow(lin_res_isgs$results)
  ngenesbeta = nrow(lin_res_genesbeta$results)
  ## results 
  isgs_lm = data.frame(lin_res_isgs$results)
  genesbeta_lm = data.frame(lin_res_genesbeta$results)
  ## xlim
  xrange = c(range(isgs_lm$Estimate)*1.3, range(genesbeta_lm$Estimate)*1.3)
  ################## summary table ###################
  ###  sig prop
  isgs.sig.no =  sum(isgs_lm$FDR < alpha)
  genesbeta.sig.no =  sum(genesbeta_lm$FDR < alpha)
  if( (isgs.sig.no >= 5) & (genesbeta.sig.no >= 5) ){
    prop.test.sig = prop.test(x = c(isgs.sig.no, genesbeta.sig.no), n = c(nisgs, ngenesbeta), correct = FALSE)
    sig.p = format(prop.test.sig$p.value, digits = 2, scientific = T)
  }else{
    prop.test.sig = prop.test(x = c(isgs.sig.no, genesbeta.sig.no), n = c(nisgs, ngenesbeta), correct = TRUE)
    sig.p = format(prop.test.sig$p.value, digits = 2, scientific = T)
  }
  ########### volcano plots ################
  if(min(isgs_lm$FDR) < alpha){
    rownames(isgs_lm) <- isgs_lm$names
    p_cutoff <- isgs_lm$p.value[min(which(isgs_lm$FDR >= alpha))]
    p1 <- ggplot(isgs_lm, aes(Estimate, -log10(p.value)) ) +
          geom_point(colour = "gray65") + 
          geom_point(data = subset(isgs_lm, FDR < alpha & Estimate > 0 ),  
             aes(Estimate, -log10(p.value)), colour="black") +
          geom_point(data = subset(isgs_lm, FDR < alpha & Estimate < 0 ),  
             aes(Estimate, -log10(p.value)), colour="gray35") +
          geom_hline(yintercept=-log10(p_cutoff), 
                     linetype="dashed", color = "black") +
          geom_vline(xintercept=0, linetype="dashed", color = "black") +
          labs(x = expression(paste("Slope" )),
               y =  expression(paste("-",log[10]," p-value" )),
               caption =  paste("Core ISGs: associations with ", clinical , sep = "") ) +
          annotate(geom="text", x = max(xrange)/1.2 ,  
                    y=-log10( p_cutoff - 0.02), 
                    label="FDR < 0.1",
                    color="black") +
          xlim( min(xrange) , max(xrange) ) + 
          theme_bw()
       
    ggsave(paste("Volcanol_Plot_ISGs_01_", cliname, sep = ""), width = 6, height = 9, device = "tiff", path = "~/Documents/gitlab/Cario_RNASeq_Microbiom_Inte/DataProcessed/interferome/plots", dpi = 300)
    print(p1)
  } else{print("No Sig.")}
  
  if(min(genesbeta_lm$FDR) < alpha){
    rownames(genesbeta_lm) <- genesbeta_lm$names
    p_cutoff2 <- genesbeta_lm$p.value[min(which(genesbeta_lm$FDR >= alpha))]
    p2 <- ggplot(genesbeta_lm, aes(Estimate, -log10(p.value)) ) +
          geom_point(colour = "gray65") + 
          geom_point(data = subset(genesbeta_lm, FDR < alpha & Estimate > 0 ),  
             aes(Estimate, -log10(p.value)), colour="black") +
          geom_point(data = subset(genesbeta_lm, FDR < alpha & Estimate < 0 ),  
             aes(Estimate, -log10(p.value)), colour="gray35") +
          geom_hline(yintercept=-log10(p_cutoff2), 
                     linetype="dashed", color = "black") +
          geom_vline(xintercept=0, linetype="dashed", color = "black") +
          labs(x = expression(paste("Slope" )),
               y =  expression(paste("-",log[10]," p-value" )),
               caption =  paste("IFN-Beta Genes: associations with ", clinical , sep = "") ) +
          annotate(geom="text", x = max(xrange)/1.2 ,  
                   y=-log10( p_cutoff2 - 0.02 ), 
                    label="FDR < 0.1",
                    color="black") +
          xlim( min(xrange) , max(xrange) ) + 
          theme_bw()
    
    ggsave(paste("Volcanol_Plot_IFNBeta_01_", cliname, sep = ""), width = 6, height = 9, device = "tiff", path = "~/Documents/gitlab/Cario_RNASeq_Microbiom_Inte/DataProcessed/interferome/plots", dpi = 300)
    print(p2)
  } else{print("No Sig.")}
  
  ## positive prop
  isgs.pos.no = sum(isgs_lm$Estimate > 0 & isgs_lm$FDR < alpha)
  genesbeta.pos.no = sum(genesbeta_lm$Estimate > 0 & genesbeta_lm$FDR < alpha)
    if(isgs.sig.no == 0 | genesbeta.sig.no == 0 ) {
       pos.p = NA
    }else{
     if( (isgs.pos.no >= 5) & (genesbeta.pos.no >= 5) ){
       prop.test.pos = prop.test(x = c(isgs.pos.no, genesbeta.pos.no), 
                              n = c(isgs.sig.no, genesbeta.sig.no), correct = FALSE)
       pos.p = format(prop.test.pos$p.value, digits = 2, scientific = T)
      }else{
       prop.test.pos = prop.test(x = c(isgs.pos.no, genesbeta.pos.no), 
                              n = c(isgs.sig.no, genesbeta.sig.no), correct = TRUE)
       pos.p = format(prop.test.pos$p.value, digits = 2, scientific = T)
      }
    }
  ## If you check prop.test, chisq.test and z-test on your data then they all give you the same p-value
  ## continuity correction if anyone less than 5
  ## the clinical summary table 
  print(clinical)
  clinical_sum[i, ] <- c(clinical, size, nisgs, ngenesbeta, isgs.sig.no,
                         round(isgs.sig.no/nisgs, 2) , genesbeta.sig.no, 
                         round(genesbeta.sig.no/ngenesbeta,2) , sig.p, isgs.pos.no,
                         round(isgs.pos.no/isgs.sig.no,2),  genesbeta.pos.no,
                         round(genesbeta.pos.no/genesbeta.sig.no,2) , 
                         pos.p)
  
  colnames(clinical_sum) <- c("Clinical Parameter", "Sample Size", "ISGs", "IFN-Beta", "No. Sig. ISGs",
                              "ISGs Sig.", "No. Sig. IFN-Betas",
                              "IFNbeta Sig.", "p value", "No. Pos. ISGs", 
                              "ISGs Pos. Prop.", "No. Pos. IFN-Betas",
                              "IFNbeta Pos. Prop.", "p value (Pos)")
}


## clinical_sum
write.xlsx(clinical_sum, 
           "~/Documents/gitlab/Cario_RNASeq_Microbiom_Inte/DataProcessed/interferome/clinical_summary_01.xlsx",
           sheetName = paste("FDR", alpha) )

clinical_sum[12:13,1] <- c("IFNalpha", "IFNbeta")
kable(data.frame(clinical_sum ) )
```

### Double Check with Normalized Counts Plot
This plot can be used as quality control.  

```{r plot_norm_counts, echo=F}

res.edger <- read.csv( "~/Documents/gitlab/Cario_RNASeq_Microbiom_Inte/DataProcessed/interferome/res.edger.csv")
cnts.edger <- read.csv("~/Documents/gitlab/Cario_RNASeq_Microbiom_Inte/DataProcessed/interferome/cnts.edger.csv",
                       row.names = 1)
## plot counts, with source code from plotCounts()
group.plot <- factor(c(rep(1,13), rep(2,19)), labels = c("Health Control", "HIV Infected"))

sim_plotcounts <- function(symbol){
  ## get genenumber
  genenumber = which(res.edger$Symbol == symbol)
  ## plot
  plot(as.integer(group.plot)+ runif(ncol(cnts.edger), -0.1, 0.1), cnts.edger[genenumber, ], 
    xlim = c(0.5, length(levels(group.plot)) + 0.8), xaxt="n" ,
    xlab = "Groups", ylab = "Normalized Counts", main = res.edger[genenumber, ]$Symbol)
axis(1, at = seq_along(levels(group.plot)), levels(group.plot))
## pos below
text(length(levels(group.plot)) + 0.5, stats::quantile( cnts.edger[genenumber, ], 0.95), 
     paste("log2 Fold Change:", round(res.edger[genenumber, ]$logFC,2),
             "","Adjusted p-value:",  
           format(res.edger[genenumber, ]$FDR,digits=2,scientific=TRUE ),"",
           "Gene_ID: ", rownames(cnts.edger)[genenumber],
                                           sep = "\n"), pos = 1)
}

symbols <- c("SH2D1A", "SCYL2", "LAG3")

for(i in symbols){
  sim_plotcounts(i)
}

```


```{r scatter_plots, eval = F, echo=F}
############# for IL-6 #######################

######### scatter plot ######################

######### viral load #####
########## ifn beta
scatter_vl_beta <- function(Symbol){
ggplot(mapping = aes(x = isgs.rld.lin[,235], 
                     y = rlog.genesbeta.05.1.raw[,which(colnames(rlog.genesbeta.05.1.raw) %in% Symbol) ])) + 
  ylab("rlog transformed counts") +
  xlab("Blood HIV Viral Load") +
  geom_point(colour = "gray15", size = 2, shape = 20) +
  geom_smooth(method = "loess", se = FALSE, colour = "steelblue") +
  geom_smooth(method = "lm", se = FALSE, colour = "tomato") +
  theme_bw() +
  theme(
    plot.title = element_text(color="black", size=14, face="bold.italic"),
    axis.title.x = element_text(color="black", size=12, face="bold"),
    axis.title.y = element_text(color="black", size=12, face="bold"),
    axis.text.x = element_text(face="bold", color="black", size=10, angle=0),
    axis.text.y = element_text(face="bold", color="black", size=10, angle=0) ) +
    annotate("text", x = mean(isgs.rld.lin[,234]), 
             y = quantile(rlog.genesbeta.05.1.raw[,which(colnames(rlog.genesbeta.05.1.raw) %in% Symbol) ], 0.95), 
             label = Symbol)

}
for( i in c("THOC3", "AQP3")){
  print(scatter_vl_beta(i))
}

##### 
lm = gene_IFNReg(isgs.05.lin, clinical_order[,6], clinical_names[1])
lm$results
sum(lm$results$Estimate > 0 & lm$results$FDR < 0.05)
sum(lm$results$FDR < 0.05)
lm$results[lm$results$FDR < 0.05, ]
format(0.0000001, digits = 4, scientific = T)


lm$results$p.value[min(which(lm$results$FDR > 0.05))]
min(which(lm$results$FDR > 0.05))
a = 1
a = a + 1
a
```